Bhavana Reddy Pothireddygari
Week 11 Assignment

In [2]:
import pandas as pd
hea_dt=pd.read_csv('HW5.csv')

In [3]:
hea_dt.head()

Name  Age  Gender Blood Type Medical Condition Date of Admission  \
0  Bobby JacksOn   30    Male         B-            Cancer        2024-01-31   
1   LesLie TErRy   62    Male         A+           Obesity        2019-08-20   
2    DaNnY sMitH   76  Female         A-           Obesity        2022-09-22   
3   andrEw waTtS   28  Female         O+          Diabetes        2020-11-18   
4  adrIENNE bEll   43  Female        AB+            Cancer        2022-09-19   

             Doctor                    Hospital Insurance Provider  \
0     Matthew Smith             Sons and Miller         Blue Cross   
1   Samantha Davies                     Kim Inc           Medicare   
2  Tiffany Mitchell                    Cook PLC              Aetna   
3       Kevin Wells  Hernandez Rogers and Vang,           Medicare   
4    Kathleen Hanna                 White-White              Aetna   

   Billing Amount  Room Number Admission Type Discharge Date   Medication  \
0    18856.281306          328         Urgent     2024-02-02  Paracetamol   
1    33643.327287          265      Emergency     2019-08-26    Ibuprofen   
2    27955.096079          205      Emergency     2022-10-07      Aspirin   
3    37909.782410          450       Elective     2020-12-18    Ibuprofen   
4    14238.317814          458         Urgent     2022-10-09   Penicillin   

   Test Results  
0        Normal  
1  Inconclusive  
2        Normal  
3      Abnormal  
4      Abnormal

In [4]:
hea_dt.dtypes

Name                   object
Age                     int64
Gender                 object
Blood Type             object
Medical Condition      object
Date of Admission      object
Doctor                 object
Hospital               object
Insurance Provider     object
Billing Amount        float64
Room Number             int64
Admission Type         object
Discharge Date         object
Medication             object
Test Results           object
dtype: object

In [5]:
print(hea_dt['Medication'].unique())


['Paracetamol' 'Ibuprofen' 'Aspirin' 'Penicillin' 'Lipitor']


In [6]:
print(hea_dt['Test Results'].unique())

['Normal' 'Inconclusive' 'Abnormal']


In [7]:
mapping = {
    'Abnormal': 1,
    'Normal': 2,
    'Inconclusive': 3 
}
hea_dt['Testing col'] = hea_dt['Test Results'].map(mapping)

In [8]:
mapping = {
    'Lipitor': 1,
    'Ibuprofen': 2,
    'Aspirin': 3,
    'Paracetamol': 4,
    'Penicillin': 5,  
}
hea_dt['Medication Numeric'] = hea_dt['Medication'].map(mapping)


In [9]:
x = hea_dt[['Age', 'Medication Numeric']]
y = hea_dt['Testing col']


In [10]:
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
import numpy as np

In [11]:
scaler = StandardScaler()
x_scaled = scaler.fit_transform(x)

In [32]:
svc = svm.SVC(kernel='linear', C=1).fit(x_scaled, y)
svc_pred = svc.predict(x_scaled)
print('SVC accuracy:', accuracy_score(y, svc_pred) * 100)

SVC accuracy: 33.800000000000004


In [33]:
# Linear SVC
lin_svc = svm.LinearSVC(C=1).fit(x_scaled, y)
lsvc_pred = lin_svc.predict(x_scaled)
print('Linear SVC accuracy:', accuracy_score(y, lsvc_pred) * 100)

C:\Users\pabbi\anaconda3\Lib\site-packages\sklearn\svm\_classes.py:31: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(


Linear SVC accuracy: 33.70990990990991


In [34]:
# RBF kernel
rbf_svc = svm.SVC(kernel='rbf', gamma=0.7, C=1).fit(x_scaled, y)
rsvc_pred = rbf_svc.predict(x_scaled)
print('RBF accuracy:', accuracy_score(y, rsvc_pred) * 100)


RBF accuracy: 34.214414414414414


In [35]:
poly_svc = svm.SVC(kernel='poly', degree=3, C=1).fit(x_scaled, y)
psvc_pred = poly_svc.predict(x_scaled)
print('Polynomial accuracy:', accuracy_score(y, psvc_pred) * 100)

Polynomial accuracy: 33.75135135135135


The Support Vector Classifier (SVC) model's accuracies, which are 33.7% for Linear SVC, 34.2% for RBF, and 33.7% for Polynomial, are all in the neighborhood of random guessing's performance level. This implies that the model might not be a good fit for the given feature set and that none of the kernel types are successfully identifying patterns in the data. This low and consistent accuracy across all kernels may be a sign that the features are not sufficiently discriminatory to allow the SVC to correctly distinguish classes, or that the data is too complicated or noisy for the SVC to handle.Consider the following actions to boost performance: hyperparameter tuning to further optimize the model, feature engineering or selection to increase the signal in the data, or even experimenting with other classification algorithms, some of which might better fit the structure of your data. Furthermore, analyzing the data distribution in greater detail and scaling aspects.



In [ ]:
import numpy as np
import matplotlib.pyplot as plt
h = 0.1 

x_min, x_max = x.iloc[:, 0].min() - 1, x.iloc[:, 0].max() + 1
y_min, y_max = x.iloc[:, 1].min() - 1, x.iloc[:, 1].max() + 1
xx, yy = np.meshgrid(np.arange(x_min, x_max, h), np.arange(y_min, y_max, h))

titles = ['linear kernel', 'linear svc', 'rbf', 'poly']

for i, clf in enumerate((svc, lin_svc, rbf_svc, poly_svc)):
    plt.subplot(2, 2, i + 1)
    plt.subplots_adjust(wspace=0.4, hspace=0.4)
    
    # Predict on the grid
    Z = clf.predict(np.c_[xx.ravel(), yy.ravel()])
    Z = Z.reshape(xx.shape)
    
    # Plot the decision boundary with contourf
    plt.contourf(xx, yy, Z, cmap=plt.cm.coolwarm, alpha=0.8)
    plt.scatter(x.iloc[:, 0], x.iloc[:, 1], c=y, cmap=plt.cm.coolwarm, edgecolor='k')
    
    # Label the axes and set limits
    plt.xlabel('Medication')
    plt.ylabel('Testing')
    plt.xlim(xx.min(), xx.max())
    plt.ylim(yy.min(), yy.max())
    plt.xticks(())
    plt.yticks(())
    plt.title(titles[i])

plt.show()


The graphs show the decision boundaries produced by a Linear Support Vector Machine (SVM) and other SVM kernels (Linear, RBF, and Polynomial) when data is categorized using two features: "Age" and "Medication Numeric." Every plot demonstrates the model's attempt to distinguish between the classes that are represented by various colored regions. The relatively straightforward, straight boundary lines displayed by the Linear Kernel and Linear SVC indicate that these models are attempting to divide the data linearly.On the other hand, more irregular decision regions that try to adjust to the data clusters are displayed by the RBF and Polynomial kernels, which have the ability to draw more intricate boundaries. The boundaries still seem a little jumbled, though, even with these more adaptable kernels, suggesting the model might be having trouble distinguishing between the classes in this feature space. This could indicate that the classes are difficult to separate, which would be consistent with the previously noted poor accuracies, or that these attributes don't offer enough differentiating power.
